lib/gollum/public/gollum/javascript/editor/langs/markdown.js

```diff
@@ -6,8 +6,35 @@
 
 (function($) {
 
-var MarkDown = $.extend(true, {}, $.DefaultLang); // The default language definition is markdown, so copy it into a new variable.
-$.GollumEditor.defineLanguage('markdown', MarkDown);
+// The default language definition (in default.js) is markdown, so we only need to set the special helper functions.
+var MarkDown = {
+
+  'gollum-helpers': {
+
+    'find-header-line': function (line, token) {
+      // For use in sections.js
+      // Checks if token matches alternative-syntax markdown headers. These are of the form '===' or '---'.
+      // If so, returns an array containing the text of the header and the previous line number.
+      // If not, returns null.
+      if ( token.match(/(^[=]+$)/) || token.match(/(^[-]+$)/) ){
+        var foundLine = line-1
+        var text = window.ace_editor.getSession().getLine(foundLine);
+        if (foundLine >= 0 && !/^\s+$/.test(text)) { // Sanity check
```

Comment:
Why the sanity check? Why assume that the previous regexp could somehow fail?

Reply:
`token` is a header found on `line`. We're looking for the title corresponding to that header, which is a different token type. It's not on the same line as `line`, that's why we're in this helper method. *If* there are situations where ACE finds a header token without there being a corresponding header title token, I thought some sanity checking might be in order: making sure at least the line exists (i.e. `line` is not 0, so `foundLine` is `>= 0`), and contains some text. But we could also leave it out: consequences would not be dire anyway.

Reply:
OK, keep as is then. 👍

Reply:
https://xkcd.com/2200/